rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2845 @@
+      {
+        "entityId": 416,
+        "properties": {
+          "http://.../text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec suscipit nisl et velit porta, eget cursus nulla fermentum. Aenean in faucibus velit, at cursus quam. Proin scelerisque quam id erat semper egestas.",
+          ...
+        }
+      },
+      // Heading Entity
+      {
+        "entityId": 417,
+        "properties": {
+          "http://.../name": "Duo Reges: constructio interrete.",
+          ...
+        }
+      },
+      // Divider Entity
+      {
+        "entityId": 418,
+        "properties": {
+          "http://.../width": "full",
+          ...
+        }
+      },
+      // User Entity
+      {
+        "entityId": 419,
+        "properties": {
+          "http://.../name": "Alice",
+          ...
+        }
+      }
+      // Page Entity
+      {
+        "entityId": 420,
+        "properties": {
+          "http://.../name": "Lorum Ipsum",
+          ...
+        },
+        "links": {
+          "http://.../written-by": 419, // referring to the User entity ID
+          "http://.../contains": [417, 416, 418] // referring to IDs of the various types of page contents above
+        }
+      }
+    ]
+    ```
+
+## JSON Schema additions
+
+The proposed type system will make use of several non-standard JSON Schema keywords to add explicitness and allow for new semantics within the schemas.
+
+The `kind` keyword will be used to identify the kind of the schema being described. The value of `kind` must be one of:
+
+- `entityType`
+- `propertyType`
+- `dataType`
+- `linkType`
+
+The `links` keyword will be used to allow Entity Type schemas to define links. The value of `links` is an object, whose keys are URIs (that point to Link Types). Values of `links` are also objects, which can optionally define extra constraints on the link (for more information see the [Reference-Level explanation](#reference-level-explanation)).
+
+The `requiredLinks` keyword will be used to specify which links are required to be present for a given Entity Type. It is a list of URIs.
+
+In JSON Schema `title` and `description` are described as follows: "Both of these keywords can be used to decorate a user interface with information about the data produced by this user interface".
+We repurpose these with slightly different meanings:
+
+The `description` keyword for Link Types adds semantic meaning to a link (which could be encapsulated in a custom vocabulary specification for the meta schemas).
+
+The `title` keyword refers to the name of the Type it is describing.
+
+The `$ref` keyword when used in the context of Entity Types, Property Types, Data Types, and Link Types URIs to specify references across the different kinds of types, the `$ref` will always need to be equal the key of the object which it is defined under. If the `$ref` is within an array definition, the URI should equal the nearest JSON Schema property name which is a URI.
```

Comment:
The other way of thinking of this is that the constraint is on _the key_, rather than on the `$ref`. It's not that users pick a key and then make the `$ref` the same – they choose the `$ref` and then set the key accordingly.

I haven't captured that here as the effect is the same, but worth thinking about in the context of validation approaches. I do think we could explain _why_ we want this restriction, and have made a suggestion to that effect.

```suggestion
The `$ref` keyword when used to refer to Property Types, Data Types, and Link Types URIs will always need to be equal the key of the object which it is defined under (and vice versa). If the `$ref` is within an array's `items` definition, the URI should equal the nearest JSON Schema property name which is a URI. This is so that the property, data, or link type can be identified from an instance conforming to it (which it could not if the key, i.e. the property name, were different from the `$ref`).
```